lib/client/mfa.go

```diff
@@ -0,0 +1,144 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/lib/auth/u2f"
+	"github.com/gravitational/teleport/lib/utils/prompt"
+)
+
+// PromptMFAChallenge prompts the user to complete MFA authentication
+// challenges.
+//
+// If promptDevicePrefix is set, it will be printed in prompts before "security
+// key" or "device". This is used to emphasize between different kinds of
+// devices, like registered vs new.
+func PromptMFAChallenge(ctx context.Context, proxyAddr string, c *proto.MFAAuthenticateChallenge, promptDevicePrefix string) (*proto.MFAAuthenticateResponse, error) {
+	if promptDevicePrefix != "" {
+		promptDevicePrefix = strings.TrimSpace(promptDevicePrefix) + " "
+	}
+
+	switch {
+	// No challenge.
+	case c.TOTP == nil && len(c.U2F) == 0:
+		return &proto.MFAAuthenticateResponse{}, nil
+	// TOTP only.
+	case c.TOTP != nil && len(c.U2F) == 0:
+		totpCode, err := prompt.Input(os.Stdout, os.Stdin, fmt.Sprintf("Enter an OTP code from a %sdevice", promptDevicePrefix))
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		return &proto.MFAAuthenticateResponse{Response: &proto.MFAAuthenticateResponse_TOTP{
+			TOTP: &proto.TOTPResponse{Code: totpCode},
+		}}, nil
+	// U2F only.
+	case c.TOTP == nil && len(c.U2F) > 0:
+		fmt.Printf("Tap any %ssecurity key\n", promptDevicePrefix)
+
+		return promptU2FChallenges(ctx, proxyAddr, c.U2F)
+	// Both TOTP and U2F.
+	case c.TOTP != nil && len(c.U2F) > 0:
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		type response struct {
+			kind string
+			resp *proto.MFAAuthenticateResponse
+			err  error
+		}
+		resCh := make(chan response)
+
+		go func() {
+			resp, err := promptU2FChallenges(ctx, proxyAddr, c.U2F)
+			select {
+			case resCh <- response{kind: "U2F", resp: resp, err: err}:
+			case <-ctx.Done():
+			}
+		}()
+
+		go func() {
+			totpCode, err := prompt.Input(os.Stdout, os.Stdin, fmt.Sprintf("Tap any %ssecurity key or enter a code from an %sOTP device", promptDevicePrefix, promptDevicePrefix))
```

Comment:
It looks the only string added in here so far is `*registered*`, which should be "a registered" rather than "an registered"?

```suggestion
			totpCode, err := prompt.Input(os.Stdout, os.Stdin, fmt.Sprintf("Tap any %ssecurity key or enter a code from a %sOTP device", promptDevicePrefix, promptDevicePrefix))
```

Reply:
Fixed message.

> Also, will you be able to read stdin (OTP) in U2F case? It looks like it could block forever.

Correct, if U2F succeeds first, this `prompt.Input` command hangs forever.
I have a TODO to support prompt cancellation: https://github.com/gravitational/teleport/blob/master/lib/utils/prompt/confirmation.go#L19-L20
It's not too bad for now, since this is for short-lived CLI commands. It only breaks when sending multiple prompts from the same process.